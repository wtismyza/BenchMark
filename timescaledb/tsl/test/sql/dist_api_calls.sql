-- This file and its contents are licensed under the Timescale License.
-- Please see the included NOTICE for copyright information and
-- LICENSE-TIMESCALE for a copy of the license.

-- Basic testing of API functions on distributed hypertable

-- Need to be super user to create extension and add data nodes
\c :TEST_DBNAME :ROLE_CLUSTER_SUPERUSER;

-- Support for execute_sql_and_filter_server_name_on_error()
\unset ECHO
\o /dev/null
\ir include/remote_exec.sql
\o
\set ECHO all

-- Cleanup from other potential tests that created these databases
SET client_min_messages TO ERROR;
DROP DATABASE IF EXISTS data_node_1;
DROP DATABASE IF EXISTS data_node_2;
DROP DATABASE IF EXISTS data_node_3;
SET client_min_messages TO NOTICE;

-- Add data nodes
SELECT * FROM add_data_node('data_node_1', host => 'localhost',
                            database => 'data_node_1');
SELECT * FROM add_data_node('data_node_2', host => 'localhost',
                            database => 'data_node_2');
SELECT * FROM add_data_node('data_node_3', host => 'localhost',
                            database => 'data_node_3');
GRANT USAGE ON FOREIGN SERVER data_node_1, data_node_2, data_node_3 TO PUBLIC;

-- Create a distributed hypertable with data
SET ROLE :ROLE_1;
CREATE TABLE disttable(
    time timestamptz NOT NULL, 
    device int, 
    value float
);
SELECT * FROM create_distributed_hypertable('disttable', 'time', 'device', 3);
INSERT INTO disttable VALUES
       ('2017-01-01 06:01', 1, 1.2),
       ('2017-01-01 09:11', 3, 4.3),
       ('2017-01-01 08:01', 1, 7.3),
       ('2017-01-02 08:01', 2, 0.23),
       ('2018-07-02 08:01', 87, 0.0),
       ('2018-07-01 06:01', 13, 3.1),
       ('2018-07-01 09:11', 90, 10303.12),
       ('2018-07-01 08:01', 29, 64);
SELECT * FROM disttable ORDER BY time;
SELECT * FROM test.remote_exec(NULL, $$ SELECT show_chunks('disttable'); $$);

-- Test APIs on the distributed hypertable

-- Call drop_chunks
SELECT drop_chunks('disttable', timestamptz '2017-03-01 00:00');
SELECT * FROM disttable ORDER BY time;
SELECT * FROM test.remote_exec(NULL, $$ SELECT show_chunks('disttable'); $$);

-- Restore the state
INSERT INTO disttable VALUES
       ('2017-01-01 06:01', 1, 1.2),
       ('2017-01-01 09:11', 3, 4.3),
       ('2017-01-01 08:01', 1, 7.3),
       ('2017-01-02 08:01', 2, 0.23);
SELECT * FROM disttable ORDER BY time;
SELECT * FROM show_chunks('disttable');
SELECT * FROM test.remote_exec(NULL, $$ SELECT show_chunks('disttable'); $$);

-- Simple test of time_bucket_gapfill, which is disabled for now.
\set ON_ERROR_STOP 0
SELECT time_bucket_gapfill('3 hours', time, '2017-01-01 06:00', '2017-01-02 18:00'),
       first(value, time),
       avg(value)
FROM disttable
GROUP BY 1;
SELECT time_bucket_gapfill('3 hours', time, '2017-01-01 06:00', '2017-01-01 18:00'),
       device,
       first(value, time),
       avg(value)
FROM disttable
GROUP BY 1,2;
SELECT
  time_bucket_gapfill('3 hours', time, '2017-01-01 06:00', '2017-01-01 18:00'),
  lag(min(time)) OVER ()
FROM disttable
GROUP BY 1;
-- Test the same queries with enabled partitionwise aggregate
SET enable_partitionwise_aggregate = 'on';
SELECT time_bucket_gapfill('3 hours', time, '2017-01-01 06:00', '2017-01-02 18:00'),
       first(value, time),
       avg(value)
FROM disttable
GROUP BY 1;
SELECT time_bucket_gapfill('3 hours', time, '2017-01-01 06:00', '2017-01-01 18:00'),
       device,
       first(value, time),
       avg(value)
FROM disttable
GROUP BY 1,2;
SET enable_partitionwise_aggregate = 'off';
\set ON_ERROR_STOP 1

-- Ensure that move_chunk() and reorder_chunk() functions cannot be used
-- with distributed hypertable
SET ROLE TO DEFAULT;
SET client_min_messages TO error;
DROP TABLESPACE IF EXISTS tablespace1;
RESET client_min_messages;
CREATE TABLESPACE tablespace1 OWNER :ROLE_CLUSTER_SUPERUSER LOCATION :TEST_TABLESPACE1_PATH;

\set ON_ERROR_STOP 0
SELECT move_chunk(chunk=>'_timescaledb_internal._dist_hyper_1_4_chunk', destination_tablespace=>'tablespace1', index_destination_tablespace=>'tablespace1', reorder_index=>'disttable_time_idx', verbose=>TRUE);
SELECT reorder_chunk('_timescaledb_internal._dist_hyper_1_4_chunk', verbose => TRUE);
\set ON_ERROR_STOP 1

DROP TABLESPACE tablespace1;

-- Ensure approximate_row_count() works with distributed hypertable
--
SELECT * FROM disttable ORDER BY time;

ANALYZE disttable;

SELECT count(*) FROM disttable;
SELECT approximate_row_count('disttable');
SELECT * FROM test.remote_exec(NULL, $$ SELECT approximate_row_count('disttable'); $$);

-- Test with native replication
--
CREATE TABLE disttable_repl(
    time timestamptz NOT NULL, 
    device int, 
    value float
);
SELECT * FROM create_distributed_hypertable('disttable_repl', 'time', 'device', 3, replication_factor => 2);
INSERT INTO disttable_repl VALUES
       ('2017-01-01 06:01', 1, 1.2),
       ('2017-01-01 09:11', 3, 4.3),
       ('2017-01-01 08:01', 1, 7.3),
       ('2017-01-02 08:01', 2, 0.23),
       ('2018-07-02 08:01', 87, 0.0),
       ('2018-07-01 06:01', 13, 3.1),
       ('2018-07-01 09:11', 90, 10303.12),
       ('2018-07-01 08:01', 29, 64);
SELECT * FROM disttable_repl ORDER BY time;

SELECT * FROM show_chunks('disttable_repl');
SELECT * FROM test.remote_exec(NULL, $$ SELECT show_chunks('disttable_repl'); $$);

SELECT count(*) FROM disttable_repl;
SELECT approximate_row_count('disttable_repl');

ANALYZE disttable_repl;
SELECT approximate_row_count('disttable_repl');

DROP TABLE disttable_repl;
