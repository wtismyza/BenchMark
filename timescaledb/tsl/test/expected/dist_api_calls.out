-- This file and its contents are licensed under the Timescale License.
-- Please see the included NOTICE for copyright information and
-- LICENSE-TIMESCALE for a copy of the license.
-- Basic testing of API functions on distributed hypertable
-- Need to be super user to create extension and add data nodes
\c :TEST_DBNAME :ROLE_CLUSTER_SUPERUSER;
-- Support for execute_sql_and_filter_server_name_on_error()
\unset ECHO
psql:include/remote_exec.sql:5: NOTICE:  schema "test" already exists, skipping
-- Cleanup from other potential tests that created these databases
SET client_min_messages TO ERROR;
DROP DATABASE IF EXISTS data_node_1;
DROP DATABASE IF EXISTS data_node_2;
DROP DATABASE IF EXISTS data_node_3;
SET client_min_messages TO NOTICE;
-- Add data nodes
SELECT * FROM add_data_node('data_node_1', host => 'localhost',
                            database => 'data_node_1');
  node_name  |   host    | port  |  database   | node_created | database_created | extension_created 
-------------+-----------+-------+-------------+--------------+------------------+-------------------
 data_node_1 | localhost | 55432 | data_node_1 | t            | t                | t
(1 row)

SELECT * FROM add_data_node('data_node_2', host => 'localhost',
                            database => 'data_node_2');
  node_name  |   host    | port  |  database   | node_created | database_created | extension_created 
-------------+-----------+-------+-------------+--------------+------------------+-------------------
 data_node_2 | localhost | 55432 | data_node_2 | t            | t                | t
(1 row)

SELECT * FROM add_data_node('data_node_3', host => 'localhost',
                            database => 'data_node_3');
  node_name  |   host    | port  |  database   | node_created | database_created | extension_created 
-------------+-----------+-------+-------------+--------------+------------------+-------------------
 data_node_3 | localhost | 55432 | data_node_3 | t            | t                | t
(1 row)

GRANT USAGE ON FOREIGN SERVER data_node_1, data_node_2, data_node_3 TO PUBLIC;
-- Create a distributed hypertable with data
SET ROLE :ROLE_1;
CREATE TABLE disttable(
    time timestamptz NOT NULL, 
    device int, 
    value float
);
SELECT * FROM create_distributed_hypertable('disttable', 'time', 'device', 3);
 hypertable_id | schema_name | table_name | created 
---------------+-------------+------------+---------
             1 | public      | disttable  | t
(1 row)

INSERT INTO disttable VALUES
       ('2017-01-01 06:01', 1, 1.2),
       ('2017-01-01 09:11', 3, 4.3),
       ('2017-01-01 08:01', 1, 7.3),
       ('2017-01-02 08:01', 2, 0.23),
       ('2018-07-02 08:01', 87, 0.0),
       ('2018-07-01 06:01', 13, 3.1),
       ('2018-07-01 09:11', 90, 10303.12),
       ('2018-07-01 08:01', 29, 64);
SELECT * FROM disttable ORDER BY time;
             time             | device |  value   
------------------------------+--------+----------
 Sun Jan 01 06:01:00 2017 PST |      1 |      1.2
 Sun Jan 01 08:01:00 2017 PST |      1 |      7.3
 Sun Jan 01 09:11:00 2017 PST |      3 |      4.3
 Mon Jan 02 08:01:00 2017 PST |      2 |     0.23
 Sun Jul 01 06:01:00 2018 PDT |     13 |      3.1
 Sun Jul 01 08:01:00 2018 PDT |     29 |       64
 Sun Jul 01 09:11:00 2018 PDT |     90 | 10303.12
 Mon Jul 02 08:01:00 2018 PDT |     87 |        0
(8 rows)

SELECT * FROM test.remote_exec(NULL, $$ SELECT show_chunks('disttable'); $$);
NOTICE:  [data_node_1]:  SELECT show_chunks('disttable')
NOTICE:  [data_node_1]:
show_chunks                                
-------------------------------------------
_timescaledb_internal._dist_hyper_1_1_chunk
_timescaledb_internal._dist_hyper_1_4_chunk
(2 rows)


NOTICE:  [data_node_2]:  SELECT show_chunks('disttable')
NOTICE:  [data_node_2]:
show_chunks                                
-------------------------------------------
_timescaledb_internal._dist_hyper_1_3_chunk
_timescaledb_internal._dist_hyper_1_5_chunk
(2 rows)


NOTICE:  [data_node_3]:  SELECT show_chunks('disttable')
NOTICE:  [data_node_3]:
show_chunks                                
-------------------------------------------
_timescaledb_internal._dist_hyper_1_2_chunk
_timescaledb_internal._dist_hyper_1_6_chunk
(2 rows)


 remote_exec 
-------------
 
(1 row)

-- Test APIs on the distributed hypertable
-- Call drop_chunks
SELECT drop_chunks('disttable', timestamptz '2017-03-01 00:00');
                 drop_chunks                 
---------------------------------------------
 _timescaledb_internal._dist_hyper_1_1_chunk
 _timescaledb_internal._dist_hyper_1_2_chunk
 _timescaledb_internal._dist_hyper_1_3_chunk
(3 rows)

SELECT * FROM disttable ORDER BY time;
             time             | device |  value   
------------------------------+--------+----------
 Sun Jul 01 06:01:00 2018 PDT |     13 |      3.1
 Sun Jul 01 08:01:00 2018 PDT |     29 |       64
 Sun Jul 01 09:11:00 2018 PDT |     90 | 10303.12
 Mon Jul 02 08:01:00 2018 PDT |     87 |        0
(4 rows)

SELECT * FROM test.remote_exec(NULL, $$ SELECT show_chunks('disttable'); $$);
NOTICE:  [data_node_1]:  SELECT show_chunks('disttable')
NOTICE:  [data_node_1]:
show_chunks                                
-------------------------------------------
_timescaledb_internal._dist_hyper_1_4_chunk
(1 row)


NOTICE:  [data_node_2]:  SELECT show_chunks('disttable')
NOTICE:  [data_node_2]:
show_chunks                                
-------------------------------------------
_timescaledb_internal._dist_hyper_1_5_chunk
(1 row)


NOTICE:  [data_node_3]:  SELECT show_chunks('disttable')
NOTICE:  [data_node_3]:
show_chunks                                
-------------------------------------------
_timescaledb_internal._dist_hyper_1_6_chunk
(1 row)


 remote_exec 
-------------
 
(1 row)

-- Restore the state
INSERT INTO disttable VALUES
       ('2017-01-01 06:01', 1, 1.2),
       ('2017-01-01 09:11', 3, 4.3),
       ('2017-01-01 08:01', 1, 7.3),
       ('2017-01-02 08:01', 2, 0.23);
SELECT * FROM disttable ORDER BY time;
             time             | device |  value   
------------------------------+--------+----------
 Sun Jan 01 06:01:00 2017 PST |      1 |      1.2
 Sun Jan 01 08:01:00 2017 PST |      1 |      7.3
 Sun Jan 01 09:11:00 2017 PST |      3 |      4.3
 Mon Jan 02 08:01:00 2017 PST |      2 |     0.23
 Sun Jul 01 06:01:00 2018 PDT |     13 |      3.1
 Sun Jul 01 08:01:00 2018 PDT |     29 |       64
 Sun Jul 01 09:11:00 2018 PDT |     90 | 10303.12
 Mon Jul 02 08:01:00 2018 PDT |     87 |        0
(8 rows)

SELECT * FROM show_chunks('disttable');
                 show_chunks                 
---------------------------------------------
 _timescaledb_internal._dist_hyper_1_4_chunk
 _timescaledb_internal._dist_hyper_1_5_chunk
 _timescaledb_internal._dist_hyper_1_6_chunk
 _timescaledb_internal._dist_hyper_1_7_chunk
 _timescaledb_internal._dist_hyper_1_8_chunk
 _timescaledb_internal._dist_hyper_1_9_chunk
(6 rows)

SELECT * FROM test.remote_exec(NULL, $$ SELECT show_chunks('disttable'); $$);
NOTICE:  [data_node_1]:  SELECT show_chunks('disttable')
NOTICE:  [data_node_1]:
show_chunks                                
-------------------------------------------
_timescaledb_internal._dist_hyper_1_4_chunk
_timescaledb_internal._dist_hyper_1_7_chunk
(2 rows)


NOTICE:  [data_node_2]:  SELECT show_chunks('disttable')
NOTICE:  [data_node_2]:
show_chunks                                
-------------------------------------------
_timescaledb_internal._dist_hyper_1_5_chunk
_timescaledb_internal._dist_hyper_1_9_chunk
(2 rows)


NOTICE:  [data_node_3]:  SELECT show_chunks('disttable')
NOTICE:  [data_node_3]:
show_chunks                                
-------------------------------------------
_timescaledb_internal._dist_hyper_1_6_chunk
_timescaledb_internal._dist_hyper_1_8_chunk
(2 rows)


 remote_exec 
-------------
 
(1 row)

-- Simple test of time_bucket_gapfill, which is disabled for now.
\set ON_ERROR_STOP 0
SELECT time_bucket_gapfill('3 hours', time, '2017-01-01 06:00', '2017-01-02 18:00'),
       first(value, time),
       avg(value)
FROM disttable
GROUP BY 1;
ERROR:  time_bucket_gapfill not implemented for distributed hypertable
SELECT time_bucket_gapfill('3 hours', time, '2017-01-01 06:00', '2017-01-01 18:00'),
       device,
       first(value, time),
       avg(value)
FROM disttable
GROUP BY 1,2;
ERROR:  time_bucket_gapfill not implemented for distributed hypertable
SELECT
  time_bucket_gapfill('3 hours', time, '2017-01-01 06:00', '2017-01-01 18:00'),
  lag(min(time)) OVER ()
FROM disttable
GROUP BY 1;
ERROR:  time_bucket_gapfill not implemented for distributed hypertable
-- Test the same queries with enabled partitionwise aggregate
SET enable_partitionwise_aggregate = 'on';
SELECT time_bucket_gapfill('3 hours', time, '2017-01-01 06:00', '2017-01-02 18:00'),
       first(value, time),
       avg(value)
FROM disttable
GROUP BY 1;
ERROR:  time_bucket_gapfill not implemented for distributed hypertable
SELECT time_bucket_gapfill('3 hours', time, '2017-01-01 06:00', '2017-01-01 18:00'),
       device,
       first(value, time),
       avg(value)
FROM disttable
GROUP BY 1,2;
ERROR:  time_bucket_gapfill not implemented for distributed hypertable
SET enable_partitionwise_aggregate = 'off';
\set ON_ERROR_STOP 1
-- Ensure that move_chunk() and reorder_chunk() functions cannot be used
-- with distributed hypertable
SET ROLE TO DEFAULT;
SET client_min_messages TO error;
DROP TABLESPACE IF EXISTS tablespace1;
RESET client_min_messages;
CREATE TABLESPACE tablespace1 OWNER :ROLE_CLUSTER_SUPERUSER LOCATION :TEST_TABLESPACE1_PATH;
\set ON_ERROR_STOP 0
SELECT move_chunk(chunk=>'_timescaledb_internal._dist_hyper_1_4_chunk', destination_tablespace=>'tablespace1', index_destination_tablespace=>'tablespace1', reorder_index=>'disttable_time_idx', verbose=>TRUE);
ERROR:  move_chunk() and reorder_chunk() cannot be used with distributed hypertables
SELECT reorder_chunk('_timescaledb_internal._dist_hyper_1_4_chunk', verbose => TRUE);
ERROR:  move_chunk() and reorder_chunk() cannot be used with distributed hypertables
\set ON_ERROR_STOP 1
DROP TABLESPACE tablespace1;
-- Ensure approximate_row_count() works with distributed hypertable
--
SELECT * FROM disttable ORDER BY time;
             time             | device |  value   
------------------------------+--------+----------
 Sun Jan 01 06:01:00 2017 PST |      1 |      1.2
 Sun Jan 01 08:01:00 2017 PST |      1 |      7.3
 Sun Jan 01 09:11:00 2017 PST |      3 |      4.3
 Mon Jan 02 08:01:00 2017 PST |      2 |     0.23
 Sun Jul 01 06:01:00 2018 PDT |     13 |      3.1
 Sun Jul 01 08:01:00 2018 PDT |     29 |       64
 Sun Jul 01 09:11:00 2018 PDT |     90 | 10303.12
 Mon Jul 02 08:01:00 2018 PDT |     87 |        0
(8 rows)

ANALYZE disttable;
SELECT count(*) FROM disttable;
 count 
-------
     8
(1 row)

SELECT approximate_row_count('disttable');
 approximate_row_count 
-----------------------
                     8
(1 row)

SELECT * FROM test.remote_exec(NULL, $$ SELECT approximate_row_count('disttable'); $$);
NOTICE:  [data_node_1]:  SELECT approximate_row_count('disttable')
NOTICE:  [data_node_1]:
approximate_row_count
---------------------
                    3
(1 row)


NOTICE:  [data_node_2]:  SELECT approximate_row_count('disttable')
NOTICE:  [data_node_2]:
approximate_row_count
---------------------
                    2
(1 row)


NOTICE:  [data_node_3]:  SELECT approximate_row_count('disttable')
NOTICE:  [data_node_3]:
approximate_row_count
---------------------
                    3
(1 row)


 remote_exec 
-------------
 
(1 row)

-- Test with native replication
--
CREATE TABLE disttable_repl(
    time timestamptz NOT NULL, 
    device int, 
    value float
);
SELECT * FROM create_distributed_hypertable('disttable_repl', 'time', 'device', 3, replication_factor => 2);
 hypertable_id | schema_name |   table_name   | created 
---------------+-------------+----------------+---------
             2 | public      | disttable_repl | t
(1 row)

INSERT INTO disttable_repl VALUES
       ('2017-01-01 06:01', 1, 1.2),
       ('2017-01-01 09:11', 3, 4.3),
       ('2017-01-01 08:01', 1, 7.3),
       ('2017-01-02 08:01', 2, 0.23),
       ('2018-07-02 08:01', 87, 0.0),
       ('2018-07-01 06:01', 13, 3.1),
       ('2018-07-01 09:11', 90, 10303.12),
       ('2018-07-01 08:01', 29, 64);
SELECT * FROM disttable_repl ORDER BY time;
             time             | device |  value   
------------------------------+--------+----------
 Sun Jan 01 06:01:00 2017 PST |      1 |      1.2
 Sun Jan 01 08:01:00 2017 PST |      1 |      7.3
 Sun Jan 01 09:11:00 2017 PST |      3 |      4.3
 Mon Jan 02 08:01:00 2017 PST |      2 |     0.23
 Sun Jul 01 06:01:00 2018 PDT |     13 |      3.1
 Sun Jul 01 08:01:00 2018 PDT |     29 |       64
 Sun Jul 01 09:11:00 2018 PDT |     90 | 10303.12
 Mon Jul 02 08:01:00 2018 PDT |     87 |        0
(8 rows)

SELECT * FROM show_chunks('disttable_repl');
                 show_chunks                  
----------------------------------------------
 _timescaledb_internal._dist_hyper_2_10_chunk
 _timescaledb_internal._dist_hyper_2_11_chunk
 _timescaledb_internal._dist_hyper_2_12_chunk
 _timescaledb_internal._dist_hyper_2_13_chunk
 _timescaledb_internal._dist_hyper_2_14_chunk
 _timescaledb_internal._dist_hyper_2_15_chunk
(6 rows)

SELECT * FROM test.remote_exec(NULL, $$ SELECT show_chunks('disttable_repl'); $$);
NOTICE:  [data_node_1]:  SELECT show_chunks('disttable_repl')
NOTICE:  [data_node_1]:
show_chunks                                 
--------------------------------------------
_timescaledb_internal._dist_hyper_2_10_chunk
_timescaledb_internal._dist_hyper_2_11_chunk
_timescaledb_internal._dist_hyper_2_13_chunk
_timescaledb_internal._dist_hyper_2_15_chunk
(4 rows)


NOTICE:  [data_node_2]:  SELECT show_chunks('disttable_repl')
NOTICE:  [data_node_2]:
show_chunks                                 
--------------------------------------------
_timescaledb_internal._dist_hyper_2_10_chunk
_timescaledb_internal._dist_hyper_2_12_chunk
_timescaledb_internal._dist_hyper_2_13_chunk
_timescaledb_internal._dist_hyper_2_14_chunk
(4 rows)


NOTICE:  [data_node_3]:  SELECT show_chunks('disttable_repl')
NOTICE:  [data_node_3]:
show_chunks                                 
--------------------------------------------
_timescaledb_internal._dist_hyper_2_11_chunk
_timescaledb_internal._dist_hyper_2_12_chunk
_timescaledb_internal._dist_hyper_2_14_chunk
_timescaledb_internal._dist_hyper_2_15_chunk
(4 rows)


 remote_exec 
-------------
 
(1 row)

SELECT count(*) FROM disttable_repl;
 count 
-------
     8
(1 row)

SELECT approximate_row_count('disttable_repl');
 approximate_row_count 
-----------------------
                     0
(1 row)

ANALYZE disttable_repl;
SELECT approximate_row_count('disttable_repl');
 approximate_row_count 
-----------------------
                     8
(1 row)

DROP TABLE disttable_repl;
